tests/conftest.py

```diff
@@ -25,6 +25,88 @@ def app():
     return app
 
 
+@pytest.fixture
+def dispatcher_local_mail_server(pytestconfig):
+    from aiosmtpd.controller import Controller
+
+    class CustomHandler:
+        async def handle_DATA(self, server, session, envelope):
+            try:
+                peer = session.peer
+                mail_from = envelope.mail_from
+                rcpt_tos = envelope.rcpt_tos
+                data = envelope.content
+                print(f"mail server: Receiving message from: {peer}")
+                print(f"mail server: Message addressed from: {mail_from}")
+                print(f"mail server: Message addressed to: {rcpt_tos}")
+                print(f"mail server: Message length : {len(data)}")
```

Comment:
here I would write to file and verify the content later in the test. Or maybe interpret from the output.

Reply:
From the context of this fixture, there is no knowledge about the session_id nor the job_id, so the scratch folder for the specific test cannot be located (maybe some information can be extracted by inspecting the content of the mail, though this wouldn't be complete). 

Any idea?

Another approach would be to just write over another file inside the test folder, and perhaps move this inside the scratch folder at the end of the test...not very elegant, but it would work...what do you think?

Reply:
> Another approach would be to just write over another file inside the test folder, and perhaps move this inside the scratch folder at the end of the test...not very elegant, but it would work...what do you think?

This file is meant to represent the events/state of the SMTP  server. Test then verifies that the SMTP server got what it should.
So there is not reason why state of the SMTP server should be in scratch directory of a particular job. 
In fact, it would not be elegant if it was.
To avoid relying on a convention on the stored filename between conftest and tests,  I'd make the fixture report it's content as an attribute.
See how caplog fixture does it.